move/axelar_gateway/sources/types/weighted_signers.move

```diff
@@ -16,21 +16,79 @@ public struct WeightedSigners has copy, drop, store {
 /// ------
 /// Invalid length of the bytes
 const EInvalidLength: u64 = 0;
+const EInvalidThreshold: u64 = 1;
+/// For when operators have changed, and proof is no longer valid.
+const EInvalidSigners: u64 = 2;
+const EMalformedSigners: u64 = 3;
+
+///-----
+/// Macros
+///-----
+
+/// Finds the first signer that satisfies the predicate.
+/// The predicate is a function that takes a `WeightedSigner` as an argument and returns a boolean.
+/// The function returns the first signer that satisfies the predicate and the index of the signer.
+/// If no signer satisfies the predicate, the error `EMalformedSigners` is raised.
+public(package) macro fun find(
+    $s: &WeightedSigners,
+    $index: u64,
+    $f: |WeightedSigner| -> bool,
+): (WeightedSigner, u64) {
+    // Bind $s to a local variable to avoid using macro parameter directly in a path expression which is not allowed
+    let s = $s;
+    let signers = s.signers();
+    let length = signers.length();
+    let mut index = $index;
+
+    // Find the first signer that satisfies the predicate
+    while (index < length && !$f(signers[index])) {
+        index = index + 1;
+    };
+
+    // If no signer satisfies the predicate, return an error
+    assert!(index < length, EMalformedSigners);
+
+    (signers[index], index)
+}
 
 /// -----------------
 /// Package Functions
 /// -----------------
+
+/// Create a new weighted signers from the given data.
 public(package) fun peel(bcs: &mut BCS): WeightedSigners {
     let len = bcs.peel_vec_length();
     assert!(len > 0, EInvalidLength);
 
     WeightedSigners {
         signers: vector::tabulate!(len, |_| weighted_signer::peel(bcs)),
         threshold: bcs.peel_u128(),
-        nonce: bytes32::peel(bcs)
+        nonce: bytes32::peel(bcs),
     }
 }
 
+/// Validates the weighted signers. The following must be true:
+/// 1. The signers are in ascending order by their public key.
+/// 2. The threshold is greater than zero.
+/// 3. The threshold is less than or equal to the total weight of the signers.
+public(package) fun validate(self: &WeightedSigners) {
+    self.validate_signers();
+    self.validate_threshold();
+}
+
+/// Finds the weight of a signer in the weighted signers by its public key.
+public(package) fun find_signer_weight(
+    self: &WeightedSigners,
+    signer_index: u64,
+    pub_key: &vector<u8>,
+): (u128, u64) {
+    // Find the signer by its public key
+    // The index of the signer is returned as well for reuse in the macro call
+    let (signer, index) = self.find!(signer_index, |signer| signer.pub_key() == pub_key);
+
+    (signer.parse_weight(), index)
```

Comment:
weight doesn't need to be validated again since it was already checked for being non-zero during signer rotation